test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,695 @@
+    free(h_buffer);
+    return ret;
+}
+
+static int check_and_free_mem(const test_params_t *params,
+                              alloc_mem_t *alloc_mem_send,
+                              alloc_mem_t *alloc_mem_recv)
+{
+    int ret;
+
+    ret = check_result(alloc_mem_recv->ptr, params->gold_data, params->size);
+    params->allocator_send->free(alloc_mem_send);
+    params->allocator_recv->free(alloc_mem_recv);
+    return ret;
+}
+
+static int test_alloc_prim_send_prim(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_no(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxPopCurrent(&primary_ctx));
+    check_no_current_context();
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    CUDA_CHECK(cuCtxPushCurrent(primary_ctx));
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_thread(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    pthread_t thread;
+    thread_arg_t thread_arg;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    thread_arg.rank     = params->rank;
+    thread_arg.send_ptr = alloc_mem_send.ptr;
+    thread_arg.recv_ptr = alloc_mem_recv.ptr;
+    thread_arg.size     = params->size;
+    pthread_create(&thread, NULL, mpi_pingpong_thread, &thread_arg);
+    pthread_join(thread, NULL);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    pop_and_release_primary_context(params->cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_user(const test_params_t *params)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx, user_ctx;
+    int res;
+
+    retain_and_push_primary_context(params->cu_dev);
+
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxPopCurrent(&primary_ctx));
+    CUDA_CHECK(cuCtxCreate(&user_ctx, 0, params->cu_dev));
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    CUDA_CHECK(cuCtxDestroy(user_ctx));
+    CUDA_CHECK(cuCtxPushCurrent(primary_ctx));
```

Comment:
can we also free with user ctx?

Reply:
Yes,  we can. But why are you asking? Would you like to change the test?
All the tests have the same pattern currently.
```
1. set context
2. allocate data
3. change context
4. transfer data
5. revert context change
6. free data
7. unset context
```

Reply:
yes, i'm afraid that we could miss some corner case, especially given that all the tests doing the same pattern

Reply:
Would you like to suggest a scenario that needs to be added?

Reply:
i was thinking of some, but seems it is mostly testing cuda rather than UCX, so no need